src/core/include/units/bits/prime.h

```diff
@@ -0,0 +1,141 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#pragma once
+
+#include <array>
+#include <cassert>
+#include <cstddef>
+#include <optional>
+
+namespace units::detail
+{
+
+constexpr bool is_prime_by_trial_division(std::size_t n) {
+  for (std::size_t f = 2; f * f <= n; f += 1 + (f % 2)) {
+    if (n % f == 0) { return false; }
+  }
+  return true;
+}
+
+constexpr std::size_t num_primes_between(std::size_t start, std::size_t end) {
+  std::size_t n = 0;
+
+  for (auto k = start; k < end; ++k) {
+    if (is_prime_by_trial_division(k)) { ++n; }
+  }
+
+  return n;
+}
+
+template<std::size_t Start, std::size_t End>
+constexpr auto primes_between() {
+  std::array<std::size_t, num_primes_between(Start, End)> results;
+  auto iter = std::begin(results);
+
+  for (auto k = Start; k < End; ++k) {
+    if (is_prime_by_trial_division(k)) {
+      *iter = k;
+      ++iter;
+    }
+  }
+
+  assert(iter == std::end(results));
+  return results;
+}
+
+// Return the first factor of n, as long as it is either k or n.
+//
+// Precondition: no integer smaller than k evenly divides n.
+constexpr std::optional<std::size_t> first_factor_maybe(std::size_t n, std::size_t k) {
+  if (n % k == 0) { return k; }
+  if (k * k > n) { return n; }
+  return std::nullopt;
+}
+
+template<std::size_t N>
+constexpr std::array<std::size_t, N> first_n_primes() {
+  std::array<std::size_t, N> primes;
+  primes[0] = 2;
+  for (std::size_t i = 1; i < N; ++i) {
+    primes[i] = primes[i - 1] + 1;
+    while (!is_prime_by_trial_division(primes[i])) { ++primes[i]; }
+  }
+  return primes;
+}
+
+template<std::size_t N>
+constexpr std::size_t product(const std::array<std::size_t, N> &values) {
+  std::size_t product = 1;
+  for (const auto &v : values) { product *= v; }
+  return product;
+}
+
+// A configurable instantiation of the "wheel factorization" algorithm [1] for prime numbers.
+//
+// Instantiate with N to use a "basis" of the first N prime numbers.  Higher values of N use fewer trial divisions, at
+// the cost of additional space.  The amount of space consumed is roughly the total number of prime numbers that are
+// less than the _product_ of the first N prime numbers.  This means it grows rapidly with N.  Consider this approximate
+// chart:
+//
+//   N | Num primes stored | Trial divisions needed ~= (stored + 1 - N) / product(basis)
+//   --+-------------------+-----------------------
+//   1 |                 1 |                 50.0 %
+//   2 |                 3 |                 33.3 %
+//   3 |                10 |                 26.7 %
+//   4 |                46 |                 20.5 %
```

Comment:
Note: this number disagrees with the figure [given on wikipedia](https://en.wikipedia.org/wiki/Wheel_factorization).  I get 43/210 for 20.5%, but they give 48/210 for 22.9%.

I think I am right---and I find that the size of `primes_in_first_wheel` for N=4 is indeed 42!  But I did want to call out that this comment differs from what one finds on wikipedia, in case anyone else sees some mistake I made.

Reply:
Actually, I think I messed up the algorithm.  I'm taking this back to draft for now.

Reply:
The root of the discrepancy is that it's not "prime numbers in the first wheel"; rather, it's "numbers coprime with the basis in the first wheel".  This means for N=4, 121 is inside the first wheel, and my current implementation would neglect it because it's not prime, but this is inappropriate because (n * 210 + 121) is **not** guaranteed to be prime for all n.

Since we're using N=3, the current version actually "works"!  But I'm not going to try to land an incorrect algorithm.  :sweat_smile: